training/CHAPEL_TUTORIAL.md

```diff
@@ -993,358 +1072,613 @@ Expected Output:
 
-There are several chapel modules available to aid in optimizing your code. The most common ones
-used by the arkouda team are [comm](https://chapel-lang.org/docs/modules/standard/CommDiagnostics.html) and
-[time](https://chapel-lang.org/docs/modules/standard/Time.html). There is also
-[memory diagnostics](https://chapel-lang.org/docs/modules/standard/Memory/Diagnostics.html), which is used less frequently.
+The biggest differences between a `coforall` and `forall` is how many tasks are created and which locales
+execute these tasks. 
+I think of `forall` loops as a reasonable default where Chapel determines this for you.
+A `coforall` offers more control, but the tradeoff is you might need to manage some stuff yourself
 
-Let's craft an example where we are intentionally copying remote values to show the
-communication that takes place.
-```Chapel
-use BlockDist, CommDiagnostics, Time;
+A `coforall` loop:
+* creates one distinct task per loop iteration, each of which executes a copy of the loop body.
+* [tasked based parallelism](https://chapel-lang.org/docs/users-guide/index.html#task-parallelism):
+I want exactly this many tasks operating in parallel
 
-var A: [Block.createDomain({0..7})] int = 0..15 by 2;
-var B: [Block.createDomain({0..15})] int = 0..15;
-writeln("A = ", A);
-writeln();
-writeln("B = ", B);
-writeln();
+A `forall` loop:
+* creates a variable number of tasks determined by the data it's iterating and the number of available processors.
+* [data based parallelism](https://chapel-lang.org/docs/users-guide/index.html#data-parallelism): operate on this
+data in parallel
 
-resetCommDiagnostics();
-startCommDiagnostics();
-var t1 = Time.timeSinceEpoch().totalSeconds();
+Here's a simple example of a `coforall`
+```Chapel
+var numTasks = 8;
 
-forall (a, i) in zip(A, A.domain) {
-  B[B.size - (2*i + 1)] = a;
+coforall tid in 1..numTasks {
+  writeln("Hello from task ", tid, " of ", numTasks);
 }
 
-var t2 = Time.timeSinceEpoch().totalSeconds() - t1;
-stopCommDiagnostics();
-writeln("Copy without aggregation time = ", t2);
-writeln();
-printCommDiagnosticsTable();
-writeln("B = ", B);
+writeln("Signing off...");
 ```
+```console
+Hello from task 4 of 8
+Hello from task 1 of 8
+Hello from task 2 of 8
+Hello from task 5 of 8
+Hello from task 3 of 8
+Hello from task 6 of 8
+Hello from task 7 of 8
+Hello from task 8 of 8
+Signing off...
 ```
-$ ./tutorial -nl 4
-A = 0 2 4 6 8 10 12 14
 
-B = 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15
+<a id="compile-multiloc"></a>
+### Enabling multiple locales
+For the rest of this tutorial, we will be using Chapel with
+[multiple locales](https://chapel-lang.org/docs/users-guide/locality/compilingAndExecutingMultiLocalePrograms.html).
+
+If you're not running on a distributed system, Chapel will simulate running in a distributed way as if you had
+multiple distinct machines. This is useful for running diagnostics and finding bugs that only show up
+in a distributed setting. 
```

Comment:
Maybe note that the performance is pretty much meaningless when running with gasnet on a non-distributed achine?